File: backend/app/schemas/dossier.py
"""Pydantic response schemas for the Person 360 dossier feature."""
from __future__ import annotations

import datetime as dt
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Nested child schemas
# ---------------------------------------------------------------------------

class DossierFamilyItem(BaseModel):
    id: int
    name: str
    relation: str
    age: Optional[int] = None
    phone: Optional[str] = None
    occupation: Optional[str] = None
    address: Optional[str] = None
    notes: Optional[str] = None

    model_config = {"from_attributes": True}


class DossierBankAccountItem(BaseModel):
    id: int
    bank_name: str
    account_no: str
    ifsc: Optional[str] = None
    branch: Optional[str] = None
    account_type: Optional[str] = None
    balance_inr: Optional[Decimal] = None
    status: Optional[str] = None
    opened_on: Optional[dt.date] = None
    flagged: Optional[bool] = False
    flag_reason: Optional[str] = None

    model_config = {"from_attributes": True}


class DossierCrimeItem(BaseModel):
    id: int
    case_ref: str
    crime_type: str
    sections: Optional[str] = None
    role: Optional[str] = None
    status: Optional[str] = None
    occurred_on: Optional[dt.date] = None
    station: Optional[str] = None
    district: Optional[str] = None
    sentence: Optional[str] = None
    narrative: Optional[str] = None

    model_config = {"from_attributes": True}


class DossierContactItem(BaseModel):
    id: int
    label: Optional[str] = None
    name: Optional[str] = None
    relation: Optional[str] = None
    phone: Optional[str] = None
    notes: Optional[str] = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# List-view item (lightweight, for the browse table)
# ---------------------------------------------------------------------------

class DossierListItem(BaseModel):
    demo_id: int
    slug: str
    full_name: str
    age: Optional[int] = None
    district: Optional[str] = None
    risk_level: Optional[str] = None
    wanted_status: Optional[str] = None
    photo_front: Optional[str] = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Detail view (full 360 dossier)
# ---------------------------------------------------------------------------

class DossierDetail(BaseModel):
    # Core identity
    demo_id: int
    slug: str
    full_name: str
    aliases: Optional[list[str]] = None
    gender: Optional[str] = None
    dob: Optional[dt.date] = None
    age: Optional[int] = None
    height_cm: Optional[int] = None
    build: Optional[str] = None
    complexion: Optional[str] = None
    identifying_marks: Optional[str] = None
    blood_group: Optional[str] = None
    nationality: Optional[str] = None
    risk_level: Optional[str] = None
    wanted_status: Optional[str] = None

    # Contact / location
    primary_phone: Optional[str] = None
    secondary_phone: Optional[str] = None
    email: Optional[str] = None
    home_address: Optional[str] = None
    district: Optional[str] = None
    pincode: Optional[str] = None

    # Photos
    photo_front: Optional[str] = None
    photo_left: Optional[str] = None
    photo_right: Optional[str] = None

    # Narrative
    summary: Optional[str] = None
    created_at: Optional[dt.datetime] = None

    # Nested relations
    family: list[DossierFamilyItem] = []
    banks: list[DossierBankAccountItem] = []
    crimes: list[DossierCrimeItem] = []
    contacts: list[DossierContactItem] = []

    # Computed aggregates
    bank_account_count: int = 0
    total_balance_inr: Decimal = Decimal("0.00")
    open_case_count: int = 0

    model_config = {"from_attributes": True}

File: frontend/src/lib/api/dossier.ts
/**
 * Person 360 dossier API client — admin-only (clearance L4+).
 * Calls /api/dossier/* which reads only demo_dossier_* tables (isolated).
 */
import { API_BASE, getAuthToken } from "./client";

async function apiFetch<T>(path: string): Promise<T> {
  const token = getAuthToken();
  const res = await fetch(`${API_BASE}${path}`, {
    headers: {
      "content-type": "application/json",
      ...(token ? { authorization: `Bearer ${token}` } : {}),
    },
  });
  if (!res.ok) throw new Error(`${path} failed: ${res.status}`);
  return res.json() as Promise<T>;
}

// ── Types ──────────────────────────────────────────────────────────────────

export type DossierListItem = {
  demo_id: number;
  slug: string;
  full_name: string;
  age: number | null;
  district: string | null;
  risk_level: string | null;
  wanted_status: string | null;
  photo_front: string | null;
};

export type DossierFamilyItem = {
  id: number;
  name: string;
  relation: string;
  age: number | null;
  phone: string | null;
  occupation: string | null;
  address: string | null;
  notes: string | null;
};

export type DossierBankAccount = {
  id: number;
  bank_name: string;
  account_no: string;
  ifsc: string | null;
  branch: string | null;
  account_type: string | null;
  balance_inr: number | null;
  status: string | null;
  opened_on: string | null;
  flagged: boolean;
  flag_reason: string | null;
};

export type DossierCrime = {
  id: number;
  case_ref: string;
  crime_type: string;
  sections: string | null;
  role: string | null;
  status: string | null;
  occurred_on: string | null;
  station: string | null;
  district: string | null;
  sentence: string | null;
  narrative: string | null;
};

export type DossierContact = {
  id: number;
  label: string | null;
  name: string | null;
  relation: string | null;
  phone: string | null;
  notes: string | null;
};

export type DossierDetail = {
  demo_id: number;
  slug: string;
  full_name: string;
  aliases: string[] | null;
  gender: string | null;
  dob: string | null;
  age: number | null;
  height_cm: number | null;
  build: string | null;
  complexion: string | null;
  identifying_marks: string | null;
  blood_group: string | null;
  nationality: string | null;
  risk_level: string | null;
  wanted_status: string | null;
  primary_phone: string | null;
  secondary_phone: string | null;
  email: string | null;
  home_address: string | null;
  district: string | null;
  pincode: string | null;
  photo_front: string | null;
  photo_left: string | null;
  photo_right: string | null;
  summary: string | null;
  created_at: string | null;
  family: DossierFamilyItem[];
  banks: DossierBankAccount[];
  crimes: DossierCrime[];
  contacts: DossierContact[];
  bank_account_count: number;
  total_balance_inr: number;
  open_case_count: number;
};

// ── API functions ──────────────────────────────────────────────────────────

export const dossier = {
  list: () => apiFetch<DossierListItem[]>("/api/dossier/list"),
  detail: (id: number) => apiFetch<DossierDetail>(`/api/dossier/${id}`),
};
